aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java

```diff
@@ -147,25 +157,50 @@ public Completable start() {
                     subscriptionProcessor.startDrainingMutationBuffer();
                 }
                 status.compareAndSet(OrchestratorStatus.STARTING, OrchestratorStatus.STARTED);
+                LOG.debug("Orchestrator started.");
             })
-        );
+        ).doFinally(startStopSemaphore::release);
         return initializationCompletable;
     }
 
     /**
      * Stop all model synchronization.
      */
-    public void stop() {
-        if (isStarted()) {
+    public synchronized void stop() {
+        boolean permitAcquired = acquirePermit(OrchestratorAction.STOP);
+        // only stop if it's started AND if we can get a permit.
+        if (!permitAcquired || !status.compareAndSet(OrchestratorStatus.STARTED, OrchestratorStatus.STOPPING)) {
+            LOG.warn(String.format("Orchestrator could not be stopped. Orchestrator status: %s", status.get()));
+            // If we acquired the permit but failed to set the status, let's release the permit.
+            if (permitAcquired) {
+                startStopSemaphore.release();
+            }
+            return;
+        }
+        try {
             LOG.info("Intentionally stopping cloud synchronization, now.");
-            status.compareAndSet(OrchestratorStatus.STARTED, OrchestratorStatus.STOPPING);
             subscriptionProcessor.stopAllSubscriptionActivity();
             storageObserver.stopObservingStorageChanges();
             mutationProcessor.stopDrainingMutationOutbox();
             status.compareAndSet(OrchestratorStatus.STOPPING, OrchestratorStatus.STOPPED);
             LOG.debug("Stopped remote synchronization.");
+        } finally {
+            startStopSemaphore.release();
         }
+    }
 
+    private boolean acquirePermit(OrchestratorAction action) {
```

Comment:
Can we wrapped the `compareAndSet` operation in the same method that grabs the semaphore so it can release it as needed?